crates/turbo-tasks-macros/src/func.rs

```diff
@@ -288,21 +304,38 @@ impl TurboFn {
             .collect()
     }
 
+    fn converted_this(&self) -> Option<Expr> {
+        self.this.as_ref().map(|Input { ty: _, ident }| {
+            parse_quote! {
+                turbo_tasks::Vc::into_raw(#ident)
+            }
+        })
+    }
+
     /// The block of the exposed function for a dynamic dispatch call to the
     /// given trait.
     pub fn dynamic_block(&self, trait_type_id_ident: &Ident) -> Block {
         let ident = &self.ident;
         let output = &self.output;
-        let converted_inputs = self.converted_inputs();
-        parse_quote! {
-            {
-                <#output as turbo_tasks::task::TaskOutput>::try_from_raw_vc(
-                    turbo_tasks::trait_call(
-                        *#trait_type_id_ident,
-                        std::borrow::Cow::Borrowed(stringify!(#ident)),
-                        vec![#converted_inputs],
+        if let Some(converted_this) = self.converted_this() {
+            let converted_inputs = self.converted_inputs();
+            parse_quote! {
+                {
+                    <#output as turbo_tasks::task::TaskOutput>::try_from_raw_vc(
+                        turbo_tasks::trait_call(
+                            *#trait_type_id_ident,
+                            std::borrow::Cow::Borrowed(stringify!(#ident)),
+                            #converted_this,
+                            vec![#converted_inputs],
+                        )
                     )
-                )
+                }
+            }
+        } else {
+            parse_quote! {
+                {
+                    unimplemented!("trait methods without self are not yet supported")
+                }
             }
         }
```

Comment:
- This might be cleaner as a `let else` statement
- What do we mean by "not yet"? Trait methods without `self` don't make sense. That's not a method, it's just an associated function (and dynamic dispatch isn't applicable)
- Maybe this should be a `compile_error!()`? https://doc.rust-lang.org/std/macro.compile_error.html